método abrirDialog (usuario decide se irá adicionar uma foto agora ou…

rubensfelipe opened the pull request:
… depois) inicializa agora na tela NovaReceitaInfo, esse método é chamado quando o usuário clica no botão salvar.

Anteriormente esse método era chamado na tela NovaReceitaFoto no método OnCreate.

Problema era que:
A dialog aparecia 2 vezes para o usuário: quando a tela NovaReceitaFoto era inicializada e logo após o usuário escolher uma foto.

rubensfelipe committed:
método abrirDialog (usuario decide se irá adicionar uma foto agora ou depois) inicializa agora na tela NovaReceitaInfo, esse método é chamado quando o usuário clica no botão salvar.

Anteriormente esse método era chamado na tela NovaReceitaFoto no método OnCreate.

Problema era que:
A dialog aparecia 2 vezes para o usuário: quando a tela NovaReceitaFoto era inicializada e logo após o usuário escolher uma foto.

rubensfelipe committed:
após clicar no botão salvar na tela NovaReceitaInfo, o teclado é encerrado.

após clicar no botão atualizar na tela EditarReceita, o teclado é encerrado.

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/NovaReceitaFotoActivity.java:
```diff
@@ -285,21 +243,70 @@ private void mensagemFotoAdicionada() {
-    private File salvarImagemDiretorio() throws IOException {
+    private void salvarImagemDirCelular(Bitmap imgFoto) throws IOException {
+
+        File filePath = Environment.getExternalStorageDirectory();
```
Existe alguma razão para você guardar o arquivo no ExternalStorageDirectory? Esse método é para guardar arquivos num cartão SD ou na memória externo, fora do domínio app. Isso só é útil para arquivos que o usuário queira usar depois. Estes dados não são deletados quando você delete o app ou quando você limpa a cache e exigem permissão de escrita. Pior, pode haver celulares que não tem acesso ao armazenamento externo e no caso deles vai dar erro na hora de tentar criar o arquivo

Você deveria usar getFilesDir(), ele te retorna um diretório dentro do sandbox do próprio app, não exige permissão de escrita e quando você limpa o cache do app os arquivos são apagados juntos. Se não me engano, usando este método, pode até apagar a permissão de escrita que você adicionou. Tente sempre usar o mínimo de permissões possíveis para o teu app

rubensfelipe replied:
pensei que era a memória interna. Vou trocar.

rubensfelipe replied:
não sei se eu li direito, mas na internet estão falando que o getExternalStorageDirectory é a memória interna para imagens.

RodrigoDavy replied:
Environment.getExternalStorageDirectory().getAbsolutePath()
Gives you the full path the SDCard. You can then do normal File I/O operations using standard Java.

Imagino que as pessoas tão sugerindo para guardar fotos no cartão SD. No caso, você também tá salvando fotos, mas são fotos que são de interesse interno do app, então é arquivo interno do app (da mesma forma que um ícone ou um thumbnail do Youtube, não é algo que você vai querer largar no cartão SD do usuário para encher memória que depois o usuário não consegue deletar com facilidade)

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/NovaReceitaFotoActivity.java:
```diff
@@ -285,21 +243,70 @@ private void mensagemFotoAdicionada() {
-    private File salvarImagemDiretorio() throws IOException {
+    private void salvarImagemDirCelular(Bitmap imgFoto) throws IOException {
+
+        File filePath = Environment.getExternalStorageDirectory();
+        File imageDir = new File(filePath.getAbsolutePath() + "/Receita Boa/" + idReceita); //caminho onde será salva a imagem na memoria interna do celular
+
+        imageDir.mkdir();
```
Essas 3 linhas poderiam ser uma função getImageDir() que retorna um imageDir com o diretório já criado

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/NovaReceitaFotoActivity.java:
```diff
@@ -285,21 +243,70 @@ private void mensagemFotoAdicionada() {
+
+        imageDir.mkdir();
+
+        String nomeImagem = idReceita + ".jpg";
+
+        File imgFile = new File(imageDir, nomeImagem);
+
+        if (!imgFile.exists()) {
+            imgFile.createNewFile(); //cria uma pasta com o arquivo a imagem dentro da galeria de fotos
+        }
```
Esta parte poderia ser uma função createImageFile()

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/NovaReceitaFotoActivity.java:
```diff
@@ -285,21 +243,70 @@ private void mensagemFotoAdicionada() {
+            imgFile.createNewFile(); //cria uma pasta com o arquivo a imagem dentro da galeria de fotos
+        }
+
+        try {
+            FileOutputStream outputStream = new FileOutputStream(imgFile);
+            imgFoto.compress(Bitmap.CompressFormat.JPEG, 75, outputStream);
+            Toast.makeText(getApplicationContext(), "Image save in internal storage", Toast.LENGTH_SHORT).show();
+
+            outputStream.flush();
+            outputStream.close();
+
+        } catch (FileNotFoundException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
```
Esta parte poderia ser uma função saveImageFile. Lembre-se de que não se deve criar uma função enorme, sempre tente separar função grandes em partes pequenas

Cada parte pequena ou função tem que ter apenas uma responsabilidade: obterDiretorio(), criarArquivo(), salvarArquivo()

Aliás, idealmente, você deveria pôr toda a lógica de salvar arquivo numa classe aparte. Uma Activity não tem por que saber como se salva um arquivo no celular, a atividade ser apenas para se comunicar com a vista :)

A vantagem também é que se você precisar criar um arquivo numa outra parte do app, você pode usar a classe ao invés de ter que repetir o código em outra atividade

rubensfelipe replied:
sim, seria uma boa criar uma nova classe. Mas eu só vou salvar imagens nessa tela.

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/NovaReceitaInfoActivity.java:
```diff
@@ -119,4 +133,49 @@ public void validarReceita(View view){ //metodo onClick (botão salvar)
+
+        //Configura titulo e mensagem
+        dialog.setTitle(getApplicationContext().getString(R.string.dialog_titulo_devo_adicionar_foto_receita));
+        dialog.setMessage(getApplicationContext().getString(R.string.dialog_msg_devo_adicionar_foto_receita));
+
+        //Configura cancelamento
+        dialog.setCancelable(false); //false: se clicar fora do alerta do dialog, o alerta não é fechado e o usuário precisa clicar em sim ou não
+
+        //Configura ações para o sim e o não
+        dialog.setPositiveButton("Sim", new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
+
+                dialog.cancel();
+
+                //Validar permissões (para acesso da camera e da galeria de fotos do usuário)
+                Permissao.validarPermissoes(permissoesNecessarias,NovaReceitaInfoActivity.this,1);
+
+                Intent i = new Intent(NovaReceitaInfoActivity.this, NovaReceitaFotoActivity.class);
+                i.putExtra("idReceita",minhasReceitas.getIdReceita());
+                startActivity(i);
+
+            }
+        });
+
+        dialog.setNegativeButton("Não", new DialogInterface.OnClickListener() {
+            @Override
+            public void onClick(DialogInterface dialog, int which) {
+
+                Toast.makeText(getApplicationContext(), getApplicationContext().getString(R.string.nova_receita_adicionada), Toast.LENGTH_SHORT).show();
+
+                finish();
+            }
+        });
+
+        //Criar e exibir AlertDialog
+        dialog.create();
+        dialog.show();
+
+    }
```
Função muito grande, separar em funções menores

rubensfelipe committed:
sugestões da PR solucionadas.

duvidas sobre o filePath.

Adição da ação de encerrar o teclado ao clicar nos botões salvar receita e
atualizar receita.

no fragment PesquisarAmigos, adição do click nos amigos e visualização da lista
das suas respectivas receitas.

Ainda não sei pq algumas imagens não querem carregar.

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/ReceitasAmigoActivity.java:
```diff
@@ -1,16 +1,184 @@
+        seguidoresRef = firebaseRef.child("seguidores"); //cria o nó seguidores (se não existir)
+        idChefLogado = UsuarioFirebaseAuth.getIdentificadorChefAuth(); //getUid do usuario logado (id do FirebaseUser)
+
+        Bundle bundle = getIntent().getExtras();
+        if (bundle != null){
+            chefAmigoSelecionado = (Chef) bundle.getSerializable("chefSelecionado");
+            idChefAmigoSelecionado = bundle.getString("idChefSeleciondado");
+
+            //Configurar referência de receitas do chef selecionado
+            receitasChefAmigoRef = ConfiguracaoFirebase.getFirebaseDatabase()
+                    .child("receitas")
+                    .child(idChefAmigoSelecionado);
+
+            //Recuperar a foto do usuário
+            String caminhoFoto = chefAmigoSelecionado.getUrlFotoChef();
+            if(caminhoFoto != null){
+                Uri url = Uri.parse(caminhoFoto); //String->Uri
+                Glide.with(ReceitasAmigoActivity.this)
+                        .load(url) //carrega a foto do FirebaseStorage através do caminho da foto
+                        .into(imagePerfil); //carrega a foto no perfil do usuário no app
+            }
```
isolar essa parte numa função (e talvez dividir a função em partes menores)

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/ReceitasAmigoActivity.java:
```diff
@@ -1,16 +1,184 @@
+
+                configurarTamanhoGrid();
+
+                //Cria uma lista de receitas do seu amigo
+                List<Receitas> receitasList = new ArrayList<>();
+                for (DataSnapshot ds: dataSnapshot.getChildren()){
+                    Receitas receita = ds.getValue(Receitas.class);
+                    receitasList.add( receita );
+                }
```
Esses blocos de código que tu põe comentário em cima são todos ótimos candidatos para virarem funções menores :)

RodrigoDavy commented on app/src/main/java/com/example/android/receitaboa/activity/ReceitasAmigoActivity.java:
```diff
@@ -1,16 +1,184 @@
+                    receitasList.add( receita );
+                }
+
+                //Configurar adapter
+                perfilAmigoAdapter = new PerfilAmigoAdapterGrid(getApplicationContext(), R.layout.grid_postagem, receitasList);
+                gridViewPerfil.setAdapter(perfilAmigoAdapter);
```
Aqui também, poderia ser uma função setAdapter()

rubensfelipe committed:
O icone de pesquisa agora tem uma animação na mainActivity.

Ao clicar, abre uma caixa de texto para digitação.

Ainda não funciona para o seu propósito.

rubensfelipe committed:
a pesquisa já está funcionando para as abas RECEITAS (minha lista de Receitas) e BUSCAR AMIGOS.

ao digitar as primeiras palavras, já aparece o resultado na pesquisa.

rubensfelipe committed:
as receitas dos seus amigos com todas as informações já podem ser visualizadas
em buscar amigos->usuario amigo->receita amigo clicada

rubensfelipe committed:
o botão de pesquisa das receitas do usuário logado na aba Receitas, estava com problemas.
Ao buscar uma receita e clicar na receita buscada, a receita selecionada era visualizada na tela VisualizarReceitaActivity.
Porém, ao voltar a MainActivity, a lista de receitas não era atualizada corretamente e ocorria um erro e o app crashava. Decidiu-se deixar essa parte para ser resolvida depois, então essa parte do código foi comentada. Logo o botão busca receitas da minha lista foi desabilitado.

Mas o botão de busca funciona normalmente na tela BuscarAmigos e foi mantido.

Adicionou-se a opção de seguir um amigo. Ao selecionar um amigo da tela Buscar Amigos,
a tela inicia com um texto do lado da foto escrito carregando. Se o usuário logado não estiver seguindo a pessoa selecionada, é setado o texto Seguir na caixa de texto.

Ao clicar no botão seguir, o texto muda para seguindo, o contador de seguidores do perfil do seu amigo é incrementando em + 1. E o contador de seguindo do perfil do usuário logado é incrementado em + 1. Exemplo: Rodrigo tem 3 seguidores e está seguindo 1 pessoa (Rubinho).
Então no seu perfil estará escrito isso. 3 seguidores 1 seguindo.

rubensfelipe committed:
Com adição da função seguir usuário, agora o usuário logado possui amigos.

As receitas dos amigos do usuário logado podem ser visualizadas na aba RECEITAS AMIGOS.
E podem ser clicadas para visualizar o modo de preparo.

rubensfelipe committed:
Ajustes na tela de login:

Aumentar o texto da TextView clicável que acessa a tela de cadastro.

Mudar o texto da TextView para o usuário saber que ele deve clicar ali para se cadastrar.

rubensfelipe committed:
reativado a pesquisa na tela minhasReceitas e ativado na tela receitas amigos.

rubensfelipe committed:
Existem alguns problemas na hora de recuperar as receitas dos amigos na aba receitas amigos.
As receitas só aparecem ao trocar de fragment, se tentar ir direto para aba receitas amigos,
a progressBar entra em um loop infinito.

A aba Feed foi atualizada, agora toda nova receita postada, desde que tenha uma foto adicionada, aparecera no Feed dos seus amigos (seguidores).
Caso vc decida por adicionar uma foto da receita quando for editar a receita, ao adicionar uma nova foto pela primeira vez, a receita com a foto será postada no Feed.

Caso o usuário tente alterar a foto da receita pela 2a vez, a foto é atualizada porém ela não será postada novamente no Feed.

Resumindo: Então só haverá postagens no Feed quando o usuário adiciona uma foto para a receita na primeira vez. Cada receita recebe apenas uma postagem no Feed.

rubensfelipe committed:
Agora as receitas que aparecem no Feed são clicáveis. Ao usuário clicar na receita do feed, ele verá todos os dados da receita (ingredientes, modo de preparo).

rubensfelipe committed:
1) A aba RECEITAS foi renomeada para aba MEU CARDÁPIO, e a aba RECEITAS AMIGOS foi renomeada para RECEITAS.
2) A aba MEUS AMIGOS mostra os amigos que vc está seguindo no app. Ao clicar em um amigo, aparecerá uma lista com as receitas dele.

rubensfelipe committed:
1) Todas as Receitas adicionadas no app agora aparecem na aba RECEITAS, elas são clicáveis e é possível pesquisar por receita.
2) Na aba MEUS AMIGOS, a pesquisa já está funcionando e os amigos são clicáveis, e ao clicar, o usuário visualiza as receitas do seu amigo.

rubensfelipe committed:
1) lista das minhas receitas na aba MEU CARDAPIO ordenada pelo nome da receita.
2) lista das receitas dos usuários do aap na aba RECEITAS ordenada pelo nome da receita.
2) lista de Amigos do usuário logado na aba MEUS AMIGOS ordenada pelo nome do amigo.

rubensfelipe committed:
lista de Usuarios cadastrados no App na aba BUSCAR AMIGOS ordenada pelo nome do usuario.

rubensfelipe committed:
1) Customizações na aba de FEED
2) Um botão de editar a receita agora está visível para o usuário, não precisa mais clicar nos 3 pontinhos

rubensfelipe committed:
1) Troca de imagens. Quando o usuário não adiciona uma foto a sua receita, aparece um frango assado, bicolor (fundo azul e frango branco).

rubensfelipe committed:
1) Foi adicionado no FEED, dentro de cada postagem, a data (contendo o dia e a hora) em que a foto foi postada pelo usuário (amigo)

rubensfelipe committed:
Adicionado um show case view na aba meu cardápio de minhas receitas fragment. Ao entrar pela primeira vez no app, um holofote mira sobre o float action button (chef) circundando-o como um feixe de luz.
Esse holofote é feito para destacar onde usuário deve clicar para adicionar a sua receita.

rubensfelipe committed:
Ajustado o showCaseView para só mostrar quando o usuário instalar o app e logar na conta dele. Um mini tutorial de como usar o app.

rubensfelipe committed:
Layout da lista das receitas alterado. Na aba RECEITAS e na aba MEU CARDAPIO.

rubensfelipe committed:
1) Inclui para ser salvo o nome do chef na aba receitas no Firebase Database para que o nome do chef possa ser mostrado na lista junto com a receita.
2) Agora na aba RECEITAS, aparece o nome do chef que postou a receita.

rubensfelipe committed:
1) Retirado a contagem de postagens do perfil do usuario. Não é necessário.

Proximos passos: apenas ajustes de design. Mas já pode ser lançado. Praticamente PRONTO.

rubensfelipe committed:
Ajuste do fundo transparente na imagem do homem da geladeira

rubensfelipe committed:
Resolvido o problema na aba MEUS AMIGOS.
PROBLEMA era: quando o usuário atualizava a sua foto de perfil, a foto não era atualizada para seus amigos, permanecendo a foto que estava no perfil do usuário no momento em que ele era seguido pelo seu amigo.
SOLUÇÃO bolada: ao atualizar a foto de perfil, a foto tb será atualizada no banco de dados para os seguidores do usuário logado.

rubensfelipe committed:
alteração do icone FAB chef para azul

rubensfelipe committed:
alteração do icone FAB chef para azul

rubensfelipe committed:
trocando o icone do app

rubensfelipe committed:
ajustado o tamanho da caixa do texto dos ingredientes. A caixa de texto aumenta de acordo com o tamanho do texto.

rubensfelipe committed:
adicionado uma dialog de carregamento ao usuário trocar a foto de perfil

rubensfelipe committed:
arrumando as progressBars de cada fragment. Enquanto a lista não é carregada, a progressBar aparece, ao carregar a lista, a progressBar some. E a progressBar não aparece se não houver nenhum item na lista.
